### ISSUE TITLE
set buffer-file-name to the repository directory name in magit mode buffers

### ISSUE THREAD

**Issue by robohack** [magit/magit#2255]
It would be extremely useful if `buffer-file-name` were set to the name of the directory of the current repository in `magit` mode buffers, perhaps in a way similar to how `pcl-cvs` does for `*cvs*` buffers.

This would make it easier to find the appropriate `magit` buffer when one has many such buffers and when their buffer names are all too long to fit enough uniquely identifying information into the standard width "Buffer" name field of a `buffer-list` buffer.  The "File" field is usually long enough and so is a good place to also show more information to help identify which buffer is which.

Sorry, I've not had time to try developing a change to do this myself yet, but I wanted to record this desire now before I forget about it again.

(And no, adjusting the width of the "Buffer" field is not a good work-around.  A balance must be struck for other uses as well since buffer lists are a global feature.)

**Comment by tarsius** [comment]
Absolutely not, that is a terrible suggestion.

This is the doc-string of `buffer-file-name`:

> **Name of file visited in current buffer, or nil if not visiting a file.** This should be an absolute file name.

Buffers whose major-modes derive from `magit-mode` do not visit files, it would therefore be wrong to set `buffer-file-name`. Doing so would completely break Magit and many other packages which have absolutely nothing to do with Magit, but which assume that other packages do not abuse `buffer-file-name`. This includes pretty much every package which in some way deals with buffers that visit files.

Instead use the buffer-local value of `default-directory` or `buffer-name`.

> And no, adjusting the width of the "Buffer" field is not a good work-around. 

It is certainly better than completely breaking Emacs.

Instead figure out to display the value of `default-directory` in the "File" column when `buffer-file-name` is nil. Shouldn't be to hard to do and there probably already is an option or package which does it.

**Comment by robohack** [comment]
Sometimes I don't think you have any idea of what makes something useable and what can make something into an absolute pain in the ass.

Nothing breaks for PCL-CVS or anything else in emacs when the directory name of `*cvs*` buffers is shown in the "File" column.  It's done it that way for _years_ and it's a "blessed" package included by default directly in emacs.

**Comment by johnmastro** [comment]
With `ibuffer` you can define your own columns via `define-ibuffer-column`.

After defining the column with whatever logic you want, you would then set `ibuffer-formats` to replace `filename-and-process` (the default "file and process" column) with your custom column.

**Comment by tarsius** [comment]
@robohack What you want is very easy to achieve:

```lisp
(add-hook 'magit-mode-hook
           (lambda ()
             (setq buffer-file-name (magit-toplevel))))
```

**Comment by johnmastro** [comment]
@robohack Or, if my `ibuffer` suggestion interests you, I posted code to do it [here](https://gist.github.com/johnmastro/c78e1d6b245440f6dab5). 

The length/complexity is from the original (i.e. not caused by adding this behavior).

**Comment by npostavs** [comment]
There seems to be a bit of confusion between interface and implementation here. I think what's actually wanted is to set `list-buffers-directory`, not `buffer-file-name`.

```
list-buffers-directory is a variable defined in `menu-bar.el'.
Its value is nil

  Automatically becomes buffer-local when set.

Documentation:
String to display in buffer listings for buffers not visiting a file.
```

**Comment by johnmastro** [comment]
Nice - and `list-buffers-directory` works with `ibuffer` too

**Comment by robohack** [comment]
Like I said -- I just want to see the directory of `magit` buffers in the `buffer-list` buffer view in the way it looks for `*cvs*` buffers.  I don't care how it works.  :-)

**Comment by tarsius** [comment]
And I never said that the suggested feature is a bad idea. I just said that the suggested implementation would break Emacs and explained why that is so. Then I also pointed toward another approach, which you ignored and instead accused me of "[not having] any idea of what makes something useable and what can make something into an absolute pain in the ass". You also made the false claim that `default-directory` is set in `*cvs*` (which I verified is not the case). Or as @npostavs put it:

> There seems to be a bit of confusion between interface and implementation here. 

Others however picked up my advice and found a ready to use feature which does exactly what you want, it just has to be activated. Which I have done now. I would have done that right away, if I had known that it was *that* easy.
